[AArch64][GlobalISel] Full reverse shuffles.

davemgreen opened the pull request:
A full shuffle reverse needs to use EXT+REV64. This adds handling for v8s16 and v16s8 types to match SDAG. Other types should be handled by perfect shuffles.

tschuett commented on llvm/lib/Target/AArch64/GISel/AArch64PostLegalizerLowering.cpp:
```diff
@@ -405,6 +405,28 @@ void applyEXT(MachineInstr &MI, ShuffleVectorPseudo &MatchInfo) {
 }
 
+bool matchFullRev(MachineInstr &MI, MachineRegisterInfo &MRI) {
+  assert(MI.getOpcode() == TargetOpcode::G_SHUFFLE_VECTOR);
```
`GShuffleVector *Shuffle = cast<GShuffleVector>(&MI);`

arsenm commented on llvm/lib/Target/AArch64/AArch64Combine.td:
```diff
@@ -131,6 +131,13 @@ def ext: GICombineRule <
+def fullrev: GICombineRule <
+  (defs root:$root, shuffle_matchdata:$matchinfo),
+  (match (G_SHUFFLE_VECTOR $src, $src1, $src2, $mask):$root,
+         [{ return matchFullRev(*${root}, MRI); }]),
```
Predicate seems like it can go in the generic combiner

davemgreen replied:
I think the types should be OK - v4i32 are either handled here or by perfect shuffles and the other types should be handled by other combines as far as I can tell.

davemgreen committed:
[AArch64] Full reverse shuffles.

A full shuffle reverse needs to use EXT+REV64. This adds handling for v8s16 and
v16s8 types to match SDAG. Other types should be handled by perfect shuffles.
